pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go

```diff
@@ -0,0 +1,380 @@
+	m.nodeShuttingDownMutex.Lock()
+	defer m.nodeShuttingDownMutex.Unlock()
+
+	if m.nodeShuttingDownNow {
+		return fmt.Errorf("node is shutting down")
+	}
+	return nil
+}
+
+func (m *managerImpl) ProcessShutdownEvent() error {
+	m.logger.V(1).Info("Shutdown manager detected new preshutdown event", "event", "preshutdown")
+
+	m.recorder.Event(m.nodeRef, v1.EventTypeNormal, kubeletevents.NodeShutdown, "Shutdown manager detected preshutdown event")
+
+	m.nodeShuttingDownMutex.Lock()
+	m.nodeShuttingDownNow = true
+	m.nodeShuttingDownMutex.Unlock()
+
+	go m.syncNodeStatus()
+
+	m.logger.V(1).Info("Shutdown manager processing preshutdown event")
+	activePods := m.getPods()
+
+	defer func() {
+		m.logger.V(1).Info("Shutdown manager completed processing preshutdown event, node will shutdown shortly")
+	}()
+
+	if m.enableMetrics && m.storage != nil {
+		startTime := time.Now()
+		err := m.storage.Store(state{
+			StartTime: startTime,
+		})
+		if err != nil {
+			m.logger.Error(err, "Failed to store graceful shutdown state")
+		}
+		metrics.GracefulShutdownStartTime.Set(timestamp(startTime))
+		metrics.GracefulShutdownEndTime.Set(0)
+
+		defer func() {
+			endTime := time.Now()
+			err := m.storage.Store(state{
+				StartTime: startTime,
+				EndTime:   endTime,
+			})
+			if err != nil {
+				m.logger.Error(err, "Failed to store graceful shutdown state")
+			}
+			metrics.GracefulShutdownEndTime.Set(timestamp(endTime))
+		}()
+	}
+
+	return m.podManager.killPods(activePods)
+}
+
+func (m *managerImpl) periodRequested() time.Duration {
+	var sum int64
+	for _, period := range m.podManager.shutdownGracePeriodByPodPriority {
+		sum += period.ShutdownGracePeriodSeconds
+	}
+	return time.Duration(sum) * time.Second
+}
+
+// Helper function to remove all occurrences of a specific item from a string list
+func removeItemFromList(stringlist []string, item string) []string {
+	writeIndex := 0
+
+	// Iterate through the list and only keep those don't match the item (case-insensitive)
+	for _, listItem := range stringlist {
+		if !strings.EqualFold(listItem, item) {
+			stringlist[writeIndex] = listItem
+			writeIndex++
+		}
+	}
+
+	// Return the modified slice, trimmed to the valid length
+	return stringlist[:writeIndex]
+}
+
+// Helper function to insert an element into a slice at a specified index
+func insertAt(slice []string, index int, value string) []string {
+	// If the index is greater than or equal to the length, append the element to the end
+	if index >= len(slice) {
+		return append(slice, value)
+	}
+
+	// Ensure there's enough space in the slice by appending a zero-value element first
+	slice = append(slice, "")
+	copy(slice[index+1:], slice[index:])
+	slice[index] = value
+
+	return slice
+}
+
+// Dependencies: ["a", "b", "c", "d"]
+// ExistingOrder: ["x", "b", "y", "c", "z"]
+// The output will be:
+// Modified List: ["x", "kubelet", "b", "y", "c", "z", "a", "d"]
+func addToExistingOrder(dependencies []string, existingOrder []string) []string {
```

Comment:
is this function reentrant?  I don't quite understand what it does, but wonder if there are any issues with it if kubelet was restarted multiple times.

Reply:
from the callsite it may look like the list will keep growing every kubelet reastart. But I might be wrong

Reply:
It will not add reduandent values. This has also been verified with e2e tests, besides the unit test cases